### PR TITLE
WIP: message list view typescript

### PR DESCRIPTION
**Pull request by evykassirer**

### CODE REVIEW

File: web/src/compose_fade.ts [hunk 1]
```diff
@@ -4,75 +4,13 @@ import assert from "minimalistic-assert";
...
+import type {MessageGroup} from "./message_list_view";
 import * as message_lists from "./message_lists";
-import type {Message} from "./message_store";
 import * as message_viewport from "./message_viewport";
 import * as people from "./people";
 import * as rows from "./rows";
-import type {TopicLink} from "./types";
 import * as util from "./util";
 
-type AllVisibilityPolicies = {
```
Comment (evykassirer) [inline web/src/compose_fade.ts:15]: move these in a commit after the typescript conversion

File: web/src/message_list_view.ts [hunk 2]
```diff
@@ -37,10 +41,92 @@ import * as sub_store from "./sub_store";
...
+    | {
+          is_stream: false;
+          is_private: true;
+          pm_with_url: string | undefined;
+          recipient_users: RecipientRowUser[];
+          display_reply_to_for_tooltip: string;
+      }
+);
+
+type MessageActions = {
```
Comment (timabbott) [inline web/src/message_list_view.ts]: `RenderingPlan` is possibly a better name for this type.

Comment (evykassirer) [reply web/src/message_list_view.ts]: Sure, changed to that!

File: web/src/message_list_view.ts [hunk 3]
```diff
@@ -270,101 +391,162 @@ function maybe_restore_focus_to_message_edit_form() {
...
-        group.invite_only = stream_data.is_invite_only_by_stream_id(
-            message_container.msg.stream_id,
-        );
-        group.is_web_public = stream_data.is_web_public(message_container.msg.stream_id);
-        group.topic = message_container.msg.topic;
-        group.match_topic = util.get_match_topic(message_container.msg);
-        group.stream_url = message_container.stream_url;
-        group.topic_url = message_container.topic_url;
-        const sub = sub_store.get(message_container.msg.stream_id);
+function populate_group_from_message(opts: {
```
Comment (evykassirer) [inline web/src/message_list_view.ts:408]: prep commit to change this function

File: web/src/message_list_view.ts [hunk 4]
```diff
@@ -270,101 +391,162 @@ function maybe_restore_focus_to_message_edit_form() {
...
+        display_reply_to_for_tooltip: message_store.get_pm_full_names(user_ids),
+    };
 }
 
+// MessageListView is the module responsible for rendering a
+// MessageList into the DOM, and maintaining it over time.
+//
+// Logic to compute context, render templates, insert them into
+// the DOM, and generally
 export class MessageListView {
```
Comment (evykassirer) [inline web/src/message_list_view.ts:499]: prep commit to move class variables up?

File: web/src/message_list_view.ts [hunk 5]
```diff
@@ -468,25 +627,43 @@ export class MessageListView {
                 edited_in_left_col: false,
                 edited_alongside_sender: false,
                 edited_status_msg: false,
+                moved: false,
+                modified: false,
```
Comment (evykassirer) [inline web/src/message_list_view.ts:631]: be part of earlier prep commit

File: web/src/message_list_view.ts [hunk 6]
```diff
@@ -570,14 +747,28 @@ export class MessageListView {
...
-            ...(mention_classname !== undefined && {mention_classname}),
+            ...(mention_classname && {mention_classname}),
             include_sender,
             ...this._maybe_get_me_message(is_hidden, message),
             ...this._get_message_edited_vars(message),
         };
     }
 
-    maybe_add_subscription_marker(group, last_msg_container, first_msg_container) {
+    maybe_add_subscription_marker_to_group(
```
Comment (evykassirer) [inline web/src/message_list_view.ts:757]: prep commit

File: web/src/message_list_view.ts [hunk 7]
```diff
@@ -587,128 +778,164 @@ export class MessageListView {
...
         }
+        return undefined;
     }
 
-    build_message_groups(message_containers) {
-        const start_group = () => ({
-            message_containers: [],
-            message_group_id: _.uniqueId("message_group_"),
-        });
+    build_message_groups(messages: Message[]): MessageGroup[] {
```
Comment (evykassirer) [inline web/src/message_list_view.ts:811]: this along with other big refactor, as prep commit

File: web/src/message_list_view.ts [hunk 8]
```diff
@@ -929,19 +1158,45 @@ export class MessageListView {
...
+            this.get_edited_notice_locations(include_sender, is_hidden, status_message),
+        );
         message_container.message_edit_notices_in_left_col = !include_sender && !is_hidden;
         message_container.message_edit_notices_alongside_sender = include_sender && !status_message;
         message_container.message_edit_notices_for_status_message =
             include_sender && status_message;
     }
 
-    render(messages, where, messages_are_new) {
+    get_edited_notice_locations(
```
Comment (evykassirer) [inline web/src/message_list_view.ts:1177]: prep for this

File: web/src/message_store.ts [hunk 9]
```diff
@@ -87,7 +87,7 @@ export const raw_message_schema = z.intersection(
         z.discriminatedUnion("type", [
             z.object({
                 type: z.literal("private"),
-                topic_links: z.optional(z.array(z.undefined())),
+                topic_links: z.optional(z.undefined()),
```
Comment (evykassirer) [inline web/src/message_store.ts:90]: this can be a commit outside of this PR

File: web/src/compose_fade.ts [hunk 10]
```diff
@@ -4,68 +4,92 @@ import assert from "minimalistic-assert";
...
 import * as compose_fade_helper from "./compose_fade_helper";
 import * as compose_state from "./compose_state";
+import type {RecipientRowUser} from "./message_list_view";
 import * as message_lists from "./message_lists";
 import type {Message} from "./message_store";
 import * as message_viewport from "./message_viewport";
 import * as people from "./people";
 import * as rows from "./rows";
 import type {TopicLink} from "./types";
-import type {AllVisibilityPolicies} from "./user_topics";
```
Comment (evykassirer) [inline web/src/compose_fade.ts:13]: this feels wrong

update `export type AllVisibilityPolicies` in user_topics

File: web/src/compose_fade.ts [hunk 11]
```diff
@@ -4,68 +4,92 @@ import assert from "minimalistic-assert";
...
 import * as util from "./util";
 
 // TODO/TypeScript: Move this to message_list_view.js when it's migrated to TypeScript.
-type MessageContainer = {
-    background_color: string;
-    date_divider_html?: string;
-    edited_alongside_sender: boolean;
-    edited_in_left_col: boolean;
-    edited_status_msg: boolean;
+export type MessageContainer = {
```
Comment (evykassirer) [inline web/src/compose_fade.ts]: clean this up to be minimal changes

File: web/src/message_lists.ts [hunk 12]
```diff
@@ -2,36 +2,10 @@ import $ from "jquery";
...
+import type {MessageListView} from "./message_list_view";
 import type {Message} from "./message_store";
 import * as ui_util from "./ui_util";
 
-// TODO(typescript): Move this to message_list_view when it's
-// converted to TypeScript.
-export type MessageContainer = {
-    msg: Message;
-    is_hidden: boolean;
-    url: string;
```
Comment (evykassirer) [inline web/src/message_lists.ts:13]: note that this was never used, not sure where `url` came from

File: web/src/message_store.ts (hunk 9 above)
Comment (evykassirer) [reply web/src/message_store.ts:90]: https://github.com/zulip/zulip/pull/31563

File: web/src/message_list_view.ts [hunk 13]
```diff
@@ -810,18 +916,15 @@ export class MessageListView {
         // prepend_groups are group to add to the bottom of the rendered DOM
         // rerender_groups are group that should be updated in place in the DOM
         // append_messages are messages which should be added to the last group in the DOM
-        // rerender_messages are messages which should be updated in place in the DOM
```
Comment (evykassirer) [inline web/src/message_list_view.ts:726]: this stale comment can be removed in a separate commit

File: web/src/message_list_view.js [hunk 14]
```diff
@@ -1011,11 +1043,8 @@ export class MessageListView {
             }
         };
 
-        if (message_containers.length === 0) {
-            return undefined;
-        }
```
Comment (evykassirer) [inline web/src/message_list_view.js:953]: code review note: This can be fully removed since we already check `message.length` earlier

File: web/src/message_list_view.js [hunk 15]
```diff
@@ -284,54 +284,84 @@ function maybe_restore_focus_to_message_edit_form() {
...
+            stream_url,
+            topic_url,
+            stream_id,
+            is_subscribed,
+            topic_is_resolved,
+            visibility_policy,
+            all_visibility_policies,
+        };
+    }
+    const user_ids = people.pm_with_user_ids(message);
```
Comment (evykassirer) [inline web/src/message_list_view.js:487]: maybe add a comment here to make it clearer this is for a private message

File: web/src/message_list_view.js [hunk 16]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
         return undefined;
     }
 
-    build_message_groups(message_containers) {
-        const start_group = () => ({
-            message_containers: [],
-            message_group_id: _.uniqueId("message_group_"),
-        });
+    build_message_groups(messages) {
+        let group_message_containers = [];
```
Comment (evykassirer) [inline web/src/message_list_view.js]: rename to `current_group_message_containers` probably, and move below `let current_group`

File: web/src/message_list_view.js [hunk 17]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
...
+            group_message_containers.push(message_container);
         };
 
-        const finish_group = () => {
-            if (current_group.message_containers.length > 0) {
-                populate_group_from_message_container(
-                    current_group,
-                    current_group.message_containers[0],
-                );
+        const finish_group = (prev_message, message_for_next_group) => {
```
Comment (evykassirer) [inline web/src/message_list_view.js]: maybe rename to `finish_group_and_start_new_group`? or separate into two functions

File: web/src/message_list_view.js [hunk 18]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
...
-            message_group_id: _.uniqueId("message_group_"),
-        });
+    build_message_groups(messages) {
+        let group_message_containers = [];
 
-        let current_group = start_group();
         const new_message_groups = [];
+        let current_group;
+
         let prev;
```
Comment (evykassirer) [inline web/src/message_list_view.js]: rename `prev` to `prev_message_container`

File: web/src/message_list_view.js [hunk 19]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
...
-                message_container.want_date_divider = date_divider_data.want_date_divider;
-                message_container.date_divider_html = date_divider_data.date_divider_html;
+                want_date_divider = date_divider_data.want_date_divider;
+                date_divider_html = date_divider_data.date_divider_html;
             } else {
-                finish_group();
-                current_group = start_group();
-                add_message_container_to_group(message_container);
-
-                update_group_date(current_group, message_container.msg, prev?.msg);
```
Comment (evykassirer) [inline web/src/message_list_view.js]: reviewer note: This went into `populate_group_from_message` in `finish_group`

File: web/src/message_list_view.js [hunk 20]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
...
-
-                message_container.include_recipient = true;
-                message_container.subscribed = false;
-                message_container.unsubscribed = false;
-
-                this.maybe_add_subscription_marker_to_group(
-                    current_group,
-                    prev?.msg,
-                    message_container.msg,
-                );
```
Comment (evykassirer) [inline web/src/message_list_view.js]: reviewer note: This went into `populate_group_from_message` in `finish_group` -- maybe I should make a `start_group` to make this clearer? And a commit where I move this code there.

File: web/src/message_list_view.js [hunk 21]
```diff
@@ -619,101 +649,119 @@ export class MessageListView {
...
-
-                if (message_container.msg.stream_id) {
```
Comment (evykassirer) [inline web/src/message_list_view.js:643]: maybe prep to change this to `type === "stream"`

File: web/src/message_list_view.ts [hunk 22]
```diff
@@ -493,6 +563,8 @@ export class MessageListView {
                 edited_in_left_col: false,
                 edited_alongside_sender: false,
                 edited_status_msg: false,
+                moved: false,
+                modified: false,
```
Comment (evykassirer) [inline web/src/message_list_view.ts:641]: prep commit to add this (plus node test change)
